### PR TITLE
feat: add log and factorial function to Int8, Int16, and Int64

### LINKED ISSUE #1246: Add log and factorial function to Int32
This ticket is related to #746.
The logarithm and factorial functions are needed in order to express the time and space complexity of certain functions.

### PR DESCRIPTION
**Pull request by KengoTODA**
Add `log2` and `factorial` functions to `Int8`, `Int16`, and `Int64` just like #1247.

I think this PR is not enough to close #1246, due to a lack of consideration for negative values. If we have a clear plan (e.g. returning 0) I will implement it in this branch later. :)

### CODE REVIEW

File: main/src/library/Int16.flix [hunk 1]
```diff
@@ -271,6 +271,18 @@ namespace Int16 {
     @Time(1) @Space(1)
     pub def flipBit(x: Int16, position: Int32): Int16 = x ^^^ (1i16 <<< position)
 
+    ///
+    /// Returns the integer binary logarithm of `x`.
+    ///
+    @Time(1) @Space(1)
+    pub def log2(x: Int16): Int32 = highestOneBitPosition(x)
```
Comment (KengoTODA) [inline main/src/library/Int16.flix:278]: It seems that other methods like `Int16.highestOneBitPosition` return an `Int32` value, so I make both `log2` and `factorial` returning `Int32` too.

File: main/src/library/Int64.flix [hunk 2]
```diff
@@ -266,6 +266,23 @@ namespace Int64 {
...
+    ///
+    /// Returns the integer binary logarithm of `x`.
+    ///
+    @Time(1) @Space(1)
+    pub def log2(x: Int64): Int32 = highestOneBitPosition(x)
+    ///
+    /// Returns the factorial of `x`.
+    ///
+    @Time(factorial(x)) @Space(1)
+    pub def factorial(x: Int64): Int64 = factorialHelper(x, 1i64)
```
Comment (KengoTODA) [inline main/src/library/Int64.flix:278]: I'm sure that it does not work for huge values like `Int64.maxValue()`.
Please tell me if there are better way to treat such values.

**Comment by magnus-madsen** [comment]
@mlutze @JonathanStarup Could either of you take a quick look?

File: main/src/library/Int64.flix [hunk 3]
```diff
@@ -266,6 +266,23 @@ namespace Int64 {
...
+
+    ///
+    /// Helper function for `factorial`.
+    ///
+    def factorialHelper(x: Int64, acc: Int64): Int64 =
```
Comment (magnus-madsen) [inline main/src/library/Int64.flix:283]: One quick comment is that we now prefer to use local defs. See e.g. how it is done in `List`.

**Comment by mlutze** [comment]
Looks reasonable to me, but two questions:
1. How does it handle negative numbers?
2. I slightly prefer keeping the result in the same type as the argument. What are your thoughts on this?

File: main/src/library/Int64.flix (hunk 2 above)
Comment (JonathanStarup) [reply main/src/library/Int64.flix:278]: I looked in java and cpp but neither seem to have factorial on the std lib, perhaps since it easily overflows.

**Comment by JonathanStarup** [comment]
> 1. How does it handle negative numbers?

It really would be nice here and many other places to have positive versions of the number types. It should probably always just return zero, exceptions or results seem too impractical.

> 2. I slightly prefer keeping the result in the same type as the argument. What are your thoughts on this?

I agree with this, log especially shouldn't be a larger type.

File: main/src/library/Int64.flix (hunk 3 above)
Comment (KengoTODA) [reply main/src/library/Int64.flix]: Thanks for your feedback, I've applied a fix as 756966e01.

**Comment by KengoTODA** [comment]
> keeping the result in the same type as the argument

I've applied a fix for `log2()` as 338b2b69e, thank you all for checking!

File: main/src/library/Int32.flix [hunk 4]
```diff
@@ -279,13 +279,12 @@ namespace Int32 {
...
     @Time(factorial(x)) @Space(1)
-    pub def factorial(x: Int32): Int32 = factorialHelper(x, 1)
-
-    ///
-    /// Helper function for `factorial`.
-    ///
-    def factorialHelper(x: Int32, acc: Int32): Int32 =
-        if (x == 0) acc else factorialHelper(x - 1, x * acc)
+    pub def factorial(x: Int32): Int32 =
+        def factorialHelper(y, acc) = match y {
```
Comment (magnus-madsen) [inline main/src/library/Int32.flix:283]: Looks good!

By convention these helper functions are now always called `loop` throughout the standard library.

**Comment by magnus-madsen** [comment]
I added one final comment and then I think this is ready for merge 👍

File: main/src/library/Int32.flix (hunk 4 above)
Comment (KengoTODA) [reply main/src/library/Int32.flix]: resolved by 60cd15b22c0b88b46b579a56a156336fe33dbc7b :+1:

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**